lambda/translcore.ml

```diff
@@ -1153,6 +1184,64 @@ and transl_match ~scopes e arg pat_expr_list partial =
     Lstaticcatch (body, (static_exception_id, val_ids), handler)
   ) classic static_handlers
 
+and prim_alloc_stack =
+  Pccall (Primitive.simple ~name:"caml_alloc_stack" ~arity:3 ~alloc:true)
+
+and transl_handler ~scopes e body val_caselist exn_caselist eff_caselist =
+  let val_fun =
+    match val_caselist with
+    | None ->
+        let param = Ident.create_local "param" in
+        lfunction ~kind:Curried ~params:[param, Pgenval]
+         ~return:Pgenval ~body:(Lvar param)
+         ~attr:default_function_attribute ~loc:Loc_unknown
+    | Some (val_caselist, partial) ->
+        let val_cases = transl_cases ~scopes val_caselist in
+        let param = Typecore.name_cases "param" val_caselist in
+        let body = Matching.for_function ~scopes e.exp_loc None (Lvar param) val_cases partial in
+        lfunction ~kind:Curried ~params:[param, Pgenval]
+          ~return:Pgenval ~attr:default_function_attribute
+          ~loc:Loc_unknown ~body
+  in
+  let exn_fun =
+    let exn_cases = transl_cases ~scopes exn_caselist in
+    let param = Typecore.name_cases "exn" exn_caselist in
+    let body = Matching.for_trywith ~scopes e.exp_loc (Lvar param) exn_cases in
+    lfunction ~kind:Curried ~params:[param, Pgenval] ~return:Pgenval
+      ~attr:default_function_attribute ~loc:Loc_unknown ~body
+  in
+  let eff_fun =
+    let param = Typecore.name_cases "eff" eff_caselist in
+    let cont = Ident.create_local "k" in
+    let cont_tail = Ident.create_local "ktail" in
+    let eff_cases = transl_cases ~scopes ~cont eff_caselist in
+    let body = Matching.for_handler ~scopes e.exp_loc (Lvar param) (Lvar cont) (Lvar cont_tail) eff_cases in
+    lfunction ~kind:Curried ~params:[(param, Pgenval); (cont, Pgenval); (cont_tail, Pgenval)]
+      ~return:Pgenval ~attr:default_function_attribute ~loc:Loc_unknown ~body
+  in
+  let is_pure = function
+    | Lconst _ -> true
+    | Lvar _ -> true
+    | Lfunction _ -> true
+    | _ -> false
```

Comment:
Should this function be moved to lambda/lambda.ml ?  Even if it has no other uses elsewhere (which I doubt), it doesn't really belong here.

Reply:
What would be an appropriate name for this function if it were moved to lambda.ml? Is `is_pure` still appropriate?

Reply:
In a lambda-calculus course, I would call this function `is_value`, because the three cases constant/variable/function abstraction are exactly the values in call-by-value for open terms...  But this  may be misleading for people who think variables are not values.

`is_pure` is a bit misleading too, as many pure expressions don't pass the test, but could be OK.

`is_simple` is somewhat vague, but conveys no intuitions, directing users to read the fine documentation comment that will say "evaluates trivially, without side effects".

Reply:
`is_evaluated`, as a less direct way of saying `is_value` but maybe dodges the potential confusion?

Reply:
Nice!  `is_evaluated` would work well indeed.

Reply:
Done in https://github.com/ocaml/ocaml/pull/12309/commits/9f83ac51d2335c20fe1172215ae9f0aba694084e.